File: Robot Control Course/Chapter8/Lesson3/Chapter8_Lesson3.java

public interface RobotRegressor {
    // returns Y(q, dq, qr, dqr, ddqr) as n x p matrix
    double[][] regressor(double[] q, double[] dq,
                         double[] qr, double[] dqr, double[] ddqr);
    int numParams();
}

public class AdaptiveCTControllerJava {
    private final int n;
    private final int p;
    private final double[][] Kd;
    private final double[][] Lambda;
    private final double[][] Gamma_d;
    private final RobotRegressor model;
    private final double[] thetaHat;

    public AdaptiveCTControllerJava(int nDof,
                                    double[][] Kd,
                                    double[][] Lambda,
                                    double[][] Gamma_d,
                                    RobotRegressor model) {
        this.n = nDof;
        this.Kd = Kd;
        this.Lambda = Lambda;
        this.Gamma_d = Gamma_d;
        this.model = model;
        this.p = model.numParams();
        this.thetaHat = new double[p];
    }

    public double[] step(double[] q, double[] dq,
                         double[] qd, double[] dqd, double[] ddqd) {
        double[] e  = sub(q, qd);
        double[] de = sub(dq, dqd);

        double[] s = add(de, matVec(Lambda, e));

        double[] dqr  = sub(dqd, matVec(Lambda, e));
        double[] ddqr = sub(ddqd, matVec(Lambda, de));

        double[][] Y = model.regressor(q, dq, dqr, dqr, ddqr);

        double[] Ytheta = matVec(Y, thetaHat);
        double[] Kds    = matVec(Kd, s);

        double[] tau = sub(Ytheta, Kds);

        double[] grad = matTVec(Y, s);
        double[] GammaGrad = matVec(Gamma_d, grad);
        for (int i = 0; i != p; ++i) {
            thetaHat[i] -= GammaGrad[i];
        }
        return tau;
    }

    // helper linear algebra (sub, add, matVec, matTVec) to be implemented
}

File: Robot Control Course/Chapter8/Lesson3/Chapter8_Lesson3.cpp

#include "Eigen/Dense"

class RobotModel {
public:
    int numParams() const;
    Eigen::MatrixXd regressor(const Eigen::VectorXd& q,
                              const Eigen::VectorXd& dq,
                              const Eigen::VectorXd& qr,
                              const Eigen::VectorXd& dqr,
                              const Eigen::VectorXd& ddqr) const;
};

class AdaptiveCTControllerCpp {
public:
    AdaptiveCTControllerCpp(int n_dof,
                            const Eigen::MatrixXd& Kd,
                            const Eigen::MatrixXd& Lambda,
                            const Eigen::MatrixXd& Gamma_d,
                            const RobotModel& model)
        : n(n_dof),
          Kd_(Kd),
          Lambda_(Lambda),
          Gamma_d_(Gamma_d),
          robot_(model)
    {
        theta_hat_.setZero(robot_.numParams());
    }

    void step(const Eigen::VectorXd& q,
              const Eigen::VectorXd& dq,
              const Eigen::VectorXd& qd,
              const Eigen::VectorXd& dqd,
              const Eigen::VectorXd& ddqd,
              Eigen::VectorXd& tau_out)
    {
        Eigen::VectorXd e  = q - qd;
        Eigen::VectorXd de = dq - dqd;
        Eigen::VectorXd s  = de + Lambda_ * e;

        Eigen::VectorXd dqr  = dqd - Lambda_ * e;
        Eigen::VectorXd ddqr = ddqd - Lambda_ * de;

        Eigen::MatrixXd Y = robot_.regressor(q, dq, dqr, ddqr);

        tau_out = Y * theta_hat_ - Kd_ * s;

        Eigen::VectorXd grad = Y.transpose() * s;
        theta_hat_ = theta_hat_ - Gamma_d_ * grad;
    }

    const Eigen::VectorXd& thetaHat() const { return theta_hat_; }

private:
    int n;
    Eigen::MatrixXd Kd_, Lambda_, Gamma_d_;
    Eigen::VectorXd theta_hat_;
    const RobotModel& robot_;
};
